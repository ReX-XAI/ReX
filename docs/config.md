# Configuration

This page describes the different sections in the ReX config file, `rex.toml`, and their possible values.
TOML is a [minimal configuration file format](https://toml.io/en/).
An example `rex.toml` file that configures a few of the possible parameters might look like this:

```none
[rex]

gpu = false

batch_size = 4

[rex.onnx]

binary_threshold = 0.5

[rex.visual]

progress_bar = true

[causal]

search_limit = 1000

iters = 30

[causal.distribution]

distribution = 'uniform'

[explanation]

[explanation.spatial]

[explanation.multi]

[explanation.evaluation]
```

N.B. that all section/subsection headers must be present even if no values in that section are set to non-default values.

## [rex] section

This section is for options that affect the general features of ReX.

mask_value = 0

Masking value for mutations, can be either an integer, float or one of the following built-in occlusions 'spectral', 'min', 'mean'.
Default: 0.

seed = 42

Random seed, only set for reproducibility.
Default: None.

gpu = true

Whether to use gpu or not.
Default: true.

batch_size = 64

Batch size for the model.
Default: 1.

### [rex.onnx] section

means = [0.485, 0.456, 0.406]

Means for min-max normalization.
Default: None.

stds = [0.229, 0.224, 0.225]

Standard deviations for min-max normalization.
Default: None.

binary_threshold = 0.5

Binary model confidence threshold.
Anything >= threshold will be classified as 1, otherwise 0.
Default: None.

norm
<!-- unclear to me what this is used for -->

Optional.
Default: None.

### [rex.visual] section

This section is for options that control the appearance of the progress bar and output data visualisation.

progress_bar = false

Whether to show progress bar in the terminal.
Defaults: true.

resize = true

Resize the explanation to the size of the original image.
This uses cubic interpolation and will not be as visually accurate as not resizing.
Default: false.

info = false

Include classification and confidence information in title of plot.
Default: true.

raw = false

Produce unvarnished image with actual masking value.
Defaults: false.

colour = 150

Pretty printing colour for explanations.
Defaults: 200.

heatmap = 'coolwarm'

Matplotlib colourscheme for responsibility map plotting.
Default: 'magma'.
See the [Matplotlib documentation](https://matplotlib.org/stable/users/explain/colors/colormaps.html) for a list of all possible colourmaps.

alpha = 0.2

Alpha blend for main image (PIL Image.blend parameter).
Default: 0.2.

grid = false

Overlay a 10*10 grid on an explanation.
Defaults: false.

mark_segments = false

Mark quickshift segmentation on image.
Default: false.

## [causal] section

This sections is for options relevant to the causal responsibility calculations.

tree_depth = 30

Maximum depth of tree.
Default: 10.
Note that search can actually go beyond this number on occasion, as the check only occurs at the end of an iteration.

search_limit = 1000

Limit on the number of combinations to consider *per iteration*.
Optional.
Default: None.
It is **not** the total work done by ReX over all iterations.
Leaving the search limit at none can potentially be very expensive.

iters = 30

Number of times to run the algorithm.
Defaults: 20.
<!-- from code looks like default is actually 30 -->

min_box_size = 10

Minimum child size, in pixels.
Default: 10.
<!-- is this width or area? -->

confidence_filter = 0.5

Remove passing mutants which have a confidence less than `confidence_filter`.
Default: 0.0 (meaning all mutants are considered).

weighted = false

Whether to weight responsibility by prediction confidence.
Default: false.

queue_style = "area"

queue_style = "intersection" | "area" | "all" | "dc"
Default: "area".

queue_len = 1

Maximum number of things to hold in search queue, either an integer or 'all'.
Default: 1.

concentrate = false

Weight responsibility by size and depth of passing partition.
Default: false.

segmentation

Default: false.

### [causal.distribution] section

This section is for options that control the distribution used to split boxes.

distribution = 'uniform'

Distribution for splitting the box.
Default: 'uniform'.
Possible choices are 'uniform' | 'binom' | 'betabinom' | 'adaptive'.

blend = 0.5

Default: 0.0.

dist_args = [1.1, 1.1]

Supplemental arguments for distribution creation, these are ignored if `distribution` does not take any parameters.
Optional.
Default: None.

## [explanation] section

This section is for options related to identifying the explanation(s).

chunk = 10

Iterate through pixel ranking in chunks, defaults to `causal.min_box_size`.

### [explanation.spatial] section

This section is for options used for spatial search for explanations.

initial_radius = 25

Initial search radius.
Default: 25.

radius_eta = 0.2

Increment to change radius.
Default: 0.2.

no_expansions = 50

Number of times to expand before quitting
Default: 1.
<!-- from code looks like default is 50? -->

### [explanation.multi] section

method = 'spotlight'
<!-- from code looks like this is not used currently -->

Multi-explanation method (only spotlight is currently implemented).

spotlights = 10

Number of spotlights to launch.
Default: 10.

spotlight_size = 24

Default size of spotlight.
Default: 20.

spotlight_eta = 0.2

Decrease spotlight by this amount.
Default: 0.2

obj_function = 'mean'

Objective function for spotlight search.
Possible options 'mean' | 'max' | 'min'.

spotlight_step

Default: 5.

### [explanation.evaluation] section

These options are used if ReX is run with the `--analyse` option.

insertion_step = 100

Insertion/deletion curve step size.
Default: 100.
