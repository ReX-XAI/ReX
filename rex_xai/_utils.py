#!/usr/bin/env python3

import importlib.metadata
from itertools import chain, combinations
from enum import Enum
from typing import Tuple, Union, Dict
from numpy.typing import NDArray
import torch as tt
import numpy as np
from skimage.segmentation import mark_boundaries
from rex_xai.logger import logger
from rex_xai.box import Box

Strategy = Enum("Strategy", ["Global", "Spatial", "MultiSpotlight", "Contrastive"])

Queue = Enum("Queue", ["Area", "All", "Intersection", "DC"])

SpatialSearch = Enum("SpatialSearch", ["NotFound", "Found"])


def powerset(r, reverse=True):
    ps = list(chain.from_iterable(combinations(r, lim) for lim in range(1, len(r) + 1)))
    if reverse:
        return reversed(ps)
    else:
        return ps


def clause_area(clause, areas: Dict) -> int:
    tot = 0
    for c in clause:
        tot += areas[c]
    return tot


class ReXError(Exception):
    pass


class ReXTomlError(ReXError):
    def __init__(self, message) -> None:
        self.message = message
        super().__init__(self.message)

    def __str__(self) -> str:
        return f"ReXTomlError: {self.message}"


class ReXPathError(ReXError):
    def __init__(self, message) -> None:
        self.message = message
        super().__init__(self.message)

    def __str__(self) -> str:
        return f"ReXPathError: no such file exists at {self.message}"


class ReXScriptError(ReXError):
    def __init__(self, message) -> None:
        self.message = message
        super().__init__(self.message)

    def __str__(self) -> str:
        return f"ReXScriptError: {self.message}"


class ReXDataError(ReXError):
    def __init__(self, message) -> None:
        self.message = message
        super().__init__(self.message)

    def __str__(self) -> str:
        return f"ReXDataError: {self.message}"


class ReXMapError(ReXError):
    def __init__(self, message) -> None:
        self.message = message
        super().__init__(self.message)

    def __str__(self) -> str:
        return f"ReXMapError: {self.message}"


def xlogx(ps):
    f = np.vectorize(_xlogx)
    return f(ps)


def _xlogx(p):
    if p == 0.0:
        return 0.0
    else:
        return p * np.log2(p)


def add_boundaries(
    img: Union[NDArray, tt.Tensor], segs: NDArray, colour=None
) -> NDArray:
    if colour is None:
        m = mark_boundaries(img, segs, mode="thick")
    else:
        m = mark_boundaries(img, segs, colour, mode="thick")
    m *= 255  # type: ignore
    m = m.astype(np.uint8)
    return m


def get_device(gpu: bool):
    if tt.backends.mps.is_available() and gpu:
        return tt.device("mps")
    if tt.device("cuda") and gpu:
        return tt.device("cuda")
    if gpu:
        logger.warning("gpu not available")
    return tt.device("cpu")


def get_map_locations(map, reverse=True):
    if isinstance(map, tt.Tensor):
        map = map.detach().cpu().numpy()
    coords = []
    for i, r in enumerate(np.nditer(map)):
        coords.append((r, np.unravel_index(i, map.shape)))
    coords = sorted(coords, reverse=reverse)
    return coords


def set_boolean_mask_value(
    tensor,
    mode,
    order,
    coords: Union[Box, Tuple[NDArray, NDArray]],
    val: bool = True,
):
    if isinstance(coords, Box):
        if mode in ("spectral", "tabular"):
            h = coords.col_start
            w = coords.col_stop
        else:
            h = slice(coords.row_start, coords.row_stop)
            w = slice(coords.col_start, coords.col_stop)
    else:
        h = coords[0]
        w = coords[1]
    # three channels
    if mode == "RGB":
        # (C, H, W)
        if order == "first":
            tensor[:, h, w] = val
        # (H, W, C)
        else:
            tensor[h, w, :] = val
    elif mode == "L":
        if order == "first":
            # (1, H, W)
            tensor[0, h, w] = val
        else:
            tensor[h, w, :] = val
    elif mode in ("spectral", "tabular"):
        if len(tensor.shape) == 1:
            tensor[h:w] = val
        else:
            tensor[0, h:w] = val
    # elif mode == "tabular":

    elif mode == "voxel":
        logger.warning("not yet implemented")


def ff(obj, fmt):
    """
    Like format(obj, fmt), but returns the string 'None' if obj is None.
    See the help for format() to see acceptable values for fmt.
    """
    return "None" if obj is None else format(obj, fmt)


def version():
    return importlib.metadata.version("rex-xai")
