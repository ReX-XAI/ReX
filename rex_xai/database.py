#!/usr/bin/env python
from datetime import datetime
import zlib
import torch as tt
import sqlalchemy as sa
from sqlalchemy import Boolean, Float, String, create_engine
from sqlalchemy import Column, Integer, Unicode
from sqlalchemy.orm import sessionmaker, DeclarativeBase
from ast import literal_eval

import pandas as pd
import numpy as np

from rex_xai.logger import logger
from rex_xai.config import CausalArgs, Strategy
from rex_xai.extraction import Explanation
from rex_xai.multi_explanation import MultiExplanation


def _dataframe(db, table):
    return pd.read_sql_table(table, f"sqlite:///{db}")


def _to_numpy(buffer, shape, dtype):
    return np.frombuffer(zlib.decompress(buffer), dtype=dtype).reshape(shape)


def db_to_pandas(db, dtype=np.float32, table="rex", process=True):
    """for interactive use"""
    df = _dataframe(db, table=table)

    if process:
        df["responsibility"] = df.apply(
            lambda row: _to_numpy(
                row["responsibility"], literal_eval(row["responsibility_shape"]), dtype
            ),
            axis=1,
        )
        #
        df["explanation"] = df.apply(
            lambda row: _to_numpy(
                row["explanation"], literal_eval(row["explanation_shape"]), np.bool_
            ),
            axis=1,
        )

    return df


def update_database(
    db,
    explanation: Explanation | MultiExplanation,  # type: ignore
    time_taken=None,
    multi=False,
):
    target_map = explanation.target_map

    if isinstance(target_map, tt.Tensor):
        target_map = target_map.detach().cpu().numpy()

    target = explanation.data.target
    if target is None:
        logger.warning("unable to update database as target is None")
        return
    classification = int(target.classification)  # type: ignore

    if not multi:
        final_mask = explanation.final_mask
        if explanation.final_mask is None:
            logger.warning("unable to update database as explanation is empty")
            return
        if isinstance(final_mask, tt.Tensor):
            final_mask = final_mask.detach().cpu().numpy()

        explanation_confidence = explanation.explanation_confidence

        add_to_database(
            db,
            explanation.args,
            classification,
            target.confidence,
            target_map,
            final_mask,
            explanation_confidence,
            time_taken,
            explanation.run_stats["total_passing"],
            explanation.run_stats["total_failing"],
            explanation.run_stats["max_depth_reached"],
            explanation.run_stats["avg_box_size"],
        )

    else:
        if type(explanation) is not MultiExplanation:
            logger.warning(
                "unable to update database, multi=True is only valid for MultiExplanation objects"
            )
            return
        else:
            for c, final_mask in enumerate(explanation.explanations):
                if isinstance(final_mask, tt.Tensor):
                    final_mask = final_mask.detach().cpu().numpy()
                add_to_database(
                    db,
                    explanation.args,
                    classification,
                    target.confidence,
                    target_map,
                    final_mask,
                    explanation.explanation_confidences[c],
                    time_taken,
                    explanation.run_stats["total_passing"],
                    explanation.run_stats["total_failing"],
                    explanation.run_stats["max_depth_reached"],
                    explanation.run_stats["avg_box_size"],
                    multi=multi,
                    multi_no=c,
                )


def add_to_database(
    db,
    args: CausalArgs,
    target,
    confidence,
    responsibility,
    explanation,
    explanation_confidence,
    time_taken,
    passing,
    failing,
    depth_reached,
    avg_box_size,
    multi=False,
    multi_no=None,
):
    if multi:
        id = hash(str(datetime.now().time()) + str(multi_no))
    else:
        id = hash(str(datetime.now().time()))

    responsibility_shape = responsibility.shape
    explanation_shape = explanation.shape

    object = DataBaseEntry(
        id,
        args.path,
        target,
        confidence,
        responsibility,
        responsibility_shape,
        explanation,
        explanation_shape,
        explanation_confidence,
        time_taken,
        depth_reached=depth_reached,
        avg_box_size=avg_box_size,
        tree_depth=args.tree_depth,
        search_limit=args.search_limit,
        iters=args.iters,
        min_size=args.min_box_size,
        distribution=str(args.distribution),
        distribution_args=str(args.distribution_args),
    )
    # if object is not None:
    object.multi = multi
    object.multi_no = multi_no
    object.passing = passing
    object.failing = failing
    object.total_work = passing + failing
    object.method = str(args.strategy)
    if args.strategy == Strategy.Spatial:
        object.spatial_radius = args.spatial_initial_radius
        object.spatial_eta = args.spatial_radius_eta
    if args.strategy == Strategy.MultiSpotlight:
        object.spotlights = args.spotlights
        object.spotlight_size = args.spotlight_size
        object.spotlight_eta = args.spotlight_eta
        object.obj_function = args.spotlight_objective_function

    db.add(object)
    db.commit()


class Base(DeclarativeBase):
    pass


class NumpyType(sa.types.TypeDecorator):
    impl = sa.types.LargeBinary

    cache_ok = True

    def process_bind_param(self, value, dialect):
        if value is not None:
            return zlib.compress(value, 9)

    def process_result_value(self, value, dialect):
        return value


class DataBaseEntry(Base):
    __tablename__ = "rex"
    id = Column(Integer, primary_key=True)
    path = Column(Unicode(100))
    target = Column(Integer)
    confidence = Column(Float)
    time = Column(Float)
    responsibility = Column(NumpyType)
    responsibility_shape = Column(Unicode)
    total_work = Column(Integer)
    passing = Column(Integer)
    failing = Column(Integer)
    explanation = Column(NumpyType)
    explanation_shape = Column(Unicode)
    explanation_confidence = Column(Float)
    multi = Column(Boolean)
    multi_no = Column(Integer)

    # causal specific columns
    depth_reached = Column(Integer)
    avg_box_size = Column(Float)
    tree_depth = Column(Integer)
    search_limit_per_iter = Column(Integer)
    iters = Column(Integer)
    min_size = Column(Integer)
    distribution = Column(String)
    distribution_args = Column(String)

    # explanation specific columns
    spatial_radius = Column(Integer)
    spatial_eta = Column(Float)

    # spotlight columns
    method = Column(String)
    spotlights = Column(Integer)
    spotlight_size = Column(Integer)
    spotlight_eta = Column(Float)
    obj_function = Column(String)

    def __init__(
        self,
        id,
        path,
        target,
        confidence,
        responsibility,
        responsibility_shape,
        explanation,
        explanation_shape,
        explanation_confidence,
        time_taken,
        passing=None,
        failing=None,
        total_work=None,
        multi=False,
        multi_no=None,
        depth_reached=None,
        avg_box_size=None,
        tree_depth=None,
        search_limit=None,
        iters=None,
        min_size=None,
        distribution=None,
        distribution_args=None,
        initial_radius=None,
        radius_eta=None,
        method=None,
        spotlights=0,
        spotlight_size=0,
        spotlight_eta=0.0,
        obj_function=None,
    ):
        self.id = id
        self.path = path
        self.target = target
        self.confidence = confidence
        self.responsibility = responsibility
        self.responsibility_shape = str(responsibility_shape)
        self.explanation = explanation
        self.explanation_shape = str(explanation_shape)
        self.explanation_confidence = explanation_confidence
        self.time = time_taken
        self.total_work = total_work
        self.passing = passing
        self.failing = failing
        # multi status
        self.multi = multi
        self.multi_no = multi_no
        # causal
        self.depth_reached = depth_reached
        self.avg_box_size = avg_box_size
        self.tree_depth = tree_depth
        self.search_limit = search_limit
        self.iters = iters
        self.min_size = min_size
        self.distribution = distribution
        self.distribution_args = distribution_args
        # spatial
        self.spatial_radius = initial_radius
        self.spatial_eta = radius_eta
        self.method = method
        # spotlights
        self.spotlights = spotlights
        self.spotlight_size = spotlight_size
        self.spotlight_eta = spotlight_eta
        self.obj_function = obj_function


def initialise_rex_db(name, echo=False):
    engine = create_engine(f"sqlite:///{name}", echo=echo)
    Base.metadata.create_all(engine, tables=[DataBaseEntry.__table__], checkfirst=True)  # type: ignore
    Session = sessionmaker(bind=engine)
    s = Session()
    return s
