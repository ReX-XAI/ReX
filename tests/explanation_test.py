import pytest

from rex_xai.config import CausalArgs, process_custom_script
from rex_xai._utils import get_device
from rex_xai.explanation import (
    try_preprocess,
    predict_target,
    get_prediction_func_from_args,
    load_and_preprocess_data,
    validate_args
)

args = CausalArgs()
args.path = "imgs/dog.jpg"
model_shape = ["N", 3, 224, 224]  # may not be correct/appropriate, check!
device = get_device(gpu=False)

process_custom_script("tests/scripts/pytorch.py", args)
data = try_preprocess(args, model_shape, device=device)
prediction_func, model_shape = get_prediction_func_from_args(args)


def test_preprocess(snapshot):
    args.custom = None
    data = load_and_preprocess_data(model_shape, device, args)
    assert data == snapshot


def test_preprocess_custom(snapshot):
    process_custom_script("tests/scripts/pytorch.py", args)
    data = load_and_preprocess_data(model_shape, device, args)
    assert data == snapshot


def test_preprocess_spectral_mask_on_image_returns_warning(caplog, snapshot):
    args.mask_value = "spectral"
    data = try_preprocess(args, model_shape, device=device)

    assert args.mask_value == 0
    assert (
        caplog.records[0].msg
        == "spectral is not suitable for images. Changing mask_value to 0"
    )
    assert data == snapshot


def test_preprocess_nii_notimplemented(caplog):
    args.path = "imgs/dog.nii"
    data = try_preprocess(args, model_shape, device=device)

    assert data == NotImplemented
    assert caplog.records[0].msg == "we do not (yet) handle nifti files generically"


def test_preprocess_npy(caplog, snapshot):
    args.path = "imgs/DoublePeakClass 0 Mean.npy"
    try_preprocess(args, model_shape, device=device)

    assert caplog.records[0].msg == "we do not generically handle this datatype"

    args.mode = "tabular"
    data = try_preprocess(args, model_shape, device=device)
    assert data == snapshot


def test_predict_target(snapshot):
    target = predict_target(data, prediction_func)

    assert target.classification == 207
    assert target.confidence == pytest.approx(0.25323787)


def test_validate_args():
    args.path = None
    with pytest.raises(FileNotFoundError):
        validate_args(args)

