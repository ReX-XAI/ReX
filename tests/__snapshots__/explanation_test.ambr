# serializer version: 1
# name: test_calculate_responsibility[Distribution.BetaBinomial-dist_args1]
  Explanation(
    args=CausalArgs(
      all=False,
      alpha=0.2,
      analyze=False,
      batch=1,
      binary_threshold=None,
      blend=0.0,
      chunk_size=10,
      colour=200,
      concentrate=False,
      confidence_filter=0.0,
      config_location=None,
      data_location=None,
      db=None,
      distribution=<Distribution.BetaBinomial: 3>,
      distribution_args=list([
        1,
        1,
      ]),
      gpu=False,
      grid=False,
      heatmap=None,
      heatmap_colours='magma',
      info=True,
      insertion_step=100,
      iters=2,
      mark_segments=False,
      mask_value=0,
      means=None,
      min_box_size=10,
      mode=None,
      model=None,
      no_expansions=50,
      norm=255.0,
      output=None,
      path='tests/test_data/dog.jpg',
      processed=False,
      progress=True,
      queue_len=1,
      queue_style=<Queue.Area: 1>,
      raw=False,
      resize=False,
      search_limit=1000,
      seed=42,
      segmentation=False,
      shape=None,
      spatial_eta=0.2,
      spatial_radius=25,
      spotlight_eta=0.2,
      spotlight_size=20,
      spotlight_step=5,
      spotlights=10,
      stds=None,
      strategy=<Strategy.Spatial: 2>,
      surface=None,
      tree_depth=10,
      verbosity=0,
      weighted=False,
    ),
    data=Data: RGB, ['N', 3, 224, 224], 224, 224, 3, first
  	 Target:CLASS: 207, CONF: 0.20462, TARGET_CLASS: None, TARGET_CONFIDENCE: None,
    explanation=None,
    final_mask=None,
    maps={207: 4},
    run_stats=dict({
      'avg_box_size': 128.16666666666669,
      'max_depth_reached': 6,
      'total_failing': 177,
      'total_passing': 89,
    }),
    target_map=array([[ 3.1666667,  4.6666665,  4.6666665, ...,  4.6666665,  4.6666665,
           4.6666665],
         [ 3.1666667,  4.6666665,  4.6666665, ...,  4.6666665,  4.6666665,
           4.6666665],
         [ 3.1666667,  4.6666665,  4.6666665, ...,  4.6666665,  4.6666665,
           4.6666665],
         ...,
         [ 6.6666665, 13.666666 , 13.666666 , ...,  5.833333 ,  5.833333 ,
           5.833333 ],
         [ 6.6666665, 13.666666 , 13.666666 , ...,  5.833333 ,  5.833333 ,
           5.833333 ],
         [ 6.6666665, 13.666666 , 13.666666 , ...,  5.833333 ,  5.833333 ,
           5.833333 ]], dtype=float32),
  )
# ---
# name: test_calculate_responsibility[Distribution.BetaBinomial-dist_args2]
  Explanation(
    args=CausalArgs(
      all=False,
      alpha=0.2,
      analyze=False,
      batch=1,
      binary_threshold=None,
      blend=0.0,
      chunk_size=10,
      colour=200,
      concentrate=False,
      confidence_filter=0.0,
      config_location=None,
      data_location=None,
      db=None,
      distribution=<Distribution.BetaBinomial: 3>,
      distribution_args=list([
        0.5,
        0.5,
      ]),
      gpu=False,
      grid=False,
      heatmap=None,
      heatmap_colours='magma',
      info=True,
      insertion_step=100,
      iters=2,
      mark_segments=False,
      mask_value=0,
      means=None,
      min_box_size=10,
      mode=None,
      model=None,
      no_expansions=50,
      norm=255.0,
      output=None,
      path='tests/test_data/dog.jpg',
      processed=False,
      progress=True,
      queue_len=1,
      queue_style=<Queue.Area: 1>,
      raw=False,
      resize=False,
      search_limit=1000,
      seed=42,
      segmentation=False,
      shape=None,
      spatial_eta=0.2,
      spatial_radius=25,
      spotlight_eta=0.2,
      spotlight_size=20,
      spotlight_step=5,
      spotlights=10,
      stds=None,
      strategy=<Strategy.Spatial: 2>,
      surface=None,
      tree_depth=10,
      verbosity=0,
      weighted=False,
    ),
    data=Data: RGB, ['N', 3, 224, 224], 224, 224, 3, first
  	 Target:CLASS: 207, CONF: 0.20462, TARGET_CLASS: None, TARGET_CONFIDENCE: None,
    explanation=None,
    final_mask=None,
    maps={207: 4},
    run_stats=dict({
      'avg_box_size': 86.0,
      'max_depth_reached': 5,
      'total_failing': 141,
      'total_passing': 83,
    }),
    target_map=array([[ 4.6666665,  4.6666665,  4.6666665, ...,  4.6666665,  4.6666665,
           4.6666665],
         [ 4.6666665,  4.6666665,  4.6666665, ...,  4.6666665,  4.6666665,
           4.6666665],
         [ 4.6666665,  4.6666665,  4.6666665, ...,  4.6666665,  4.6666665,
           4.6666665],
         ...,
         [16.       , 16.       , 16.       , ..., 12.833334 ,  5.833333 ,
           5.833333 ],
         [16.       , 16.       , 16.       , ..., 12.833334 ,  5.833333 ,
           5.833333 ],
         [16.       , 16.       , 16.       , ..., 12.833334 ,  5.833333 ,
           5.833333 ]], dtype=float32),
  )
# ---
# name: test_calculate_responsibility[Distribution.BetaBinomial-dist_args3]
  Explanation(
    args=CausalArgs(
      all=False,
      alpha=0.2,
      analyze=False,
      batch=1,
      binary_threshold=None,
      blend=0.0,
      chunk_size=10,
      colour=200,
      concentrate=False,
      confidence_filter=0.0,
      config_location=None,
      data_location=None,
      db=None,
      distribution=<Distribution.BetaBinomial: 3>,
      distribution_args=list([
        1,
        0.5,
      ]),
      gpu=False,
      grid=False,
      heatmap=None,
      heatmap_colours='magma',
      info=True,
      insertion_step=100,
      iters=2,
      mark_segments=False,
      mask_value=0,
      means=None,
      min_box_size=10,
      mode=None,
      model=None,
      no_expansions=50,
      norm=255.0,
      output=None,
      path='tests/test_data/dog.jpg',
      processed=False,
      progress=True,
      queue_len=1,
      queue_style=<Queue.Area: 1>,
      raw=False,
      resize=False,
      search_limit=1000,
      seed=42,
      segmentation=False,
      shape=None,
      spatial_eta=0.2,
      spatial_radius=25,
      spotlight_eta=0.2,
      spotlight_size=20,
      spotlight_step=5,
      spotlights=10,
      stds=None,
      strategy=<Strategy.Spatial: 2>,
      surface=None,
      tree_depth=10,
      verbosity=0,
      weighted=False,
    ),
    data=Data: RGB, ['N', 3, 224, 224], 224, 224, 3, first
  	 Target:CLASS: 207, CONF: 0.20462, TARGET_CLASS: None, TARGET_CONFIDENCE: None,
    explanation=None,
    final_mask=None,
    maps={207: 4},
    run_stats=dict({
      'avg_box_size': 215.25,
      'max_depth_reached': 6,
      'total_failing': 102,
      'total_passing': 66,
    }),
    target_map=array([[4.6666665, 4.6666665, 4.6666665, ..., 4.6666665, 4.6666665,
          4.6666665],
         [4.6666665, 4.6666665, 4.6666665, ..., 4.6666665, 4.6666665,
          4.6666665],
         [4.6666665, 4.6666665, 4.6666665, ..., 4.6666665, 4.6666665,
          4.6666665],
         ...,
         [5.666667 , 5.666667 , 5.666667 , ..., 8.5      , 8.5      ,
          8.5      ],
         [5.666667 , 5.666667 , 5.666667 , ..., 8.5      , 8.5      ,
          8.5      ],
         [5.666667 , 5.666667 , 5.666667 , ..., 8.5      , 8.5      ,
          8.5      ]], dtype=float32),
  )
# ---
# name: test_calculate_responsibility[Distribution.BetaBinomial-dist_args4]
  Explanation(
    args=CausalArgs(
      all=False,
      alpha=0.2,
      analyze=False,
      batch=1,
      binary_threshold=None,
      blend=0.0,
      chunk_size=10,
      colour=200,
      concentrate=False,
      confidence_filter=0.0,
      config_location=None,
      data_location=None,
      db=None,
      distribution=<Distribution.BetaBinomial: 3>,
      distribution_args=list([
        0.5,
        1,
      ]),
      gpu=False,
      grid=False,
      heatmap=None,
      heatmap_colours='magma',
      info=True,
      insertion_step=100,
      iters=2,
      mark_segments=False,
      mask_value=0,
      means=None,
      min_box_size=10,
      mode=None,
      model=None,
      no_expansions=50,
      norm=255.0,
      output=None,
      path='tests/test_data/dog.jpg',
      processed=False,
      progress=True,
      queue_len=1,
      queue_style=<Queue.Area: 1>,
      raw=False,
      resize=False,
      search_limit=1000,
      seed=42,
      segmentation=False,
      shape=None,
      spatial_eta=0.2,
      spatial_radius=25,
      spotlight_eta=0.2,
      spotlight_size=20,
      spotlight_step=5,
      spotlights=10,
      stds=None,
      strategy=<Strategy.Spatial: 2>,
      surface=None,
      tree_depth=10,
      verbosity=0,
      weighted=False,
    ),
    data=Data: RGB, ['N', 3, 224, 224], 224, 224, 3, first
  	 Target:CLASS: 207, CONF: 0.20462, TARGET_CLASS: None, TARGET_CONFIDENCE: None,
    explanation=None,
    final_mask=None,
    maps={207: 4},
    run_stats=dict({
      'avg_box_size': 850.3333333333334,
      'max_depth_reached': 7,
      'total_failing': 157,
      'total_passing': 81,
    }),
    target_map=array([[ 2.3333335,  2.3333335,  2.3333335, ...,  2.3333335,  2.3333335,
           2.3333335],
         [ 2.3333335,  2.3333335,  2.3333335, ...,  2.3333335,  2.3333335,
           2.3333335],
         [ 2.3333335,  2.3333335,  2.3333335, ...,  2.3333335,  2.3333335,
           2.3333335],
         ...,
         [ 2.3333335,  2.3333335,  2.3333335, ..., 11.666667 , 11.666667 ,
          11.666667 ],
         [ 2.3333335,  2.3333335,  2.3333335, ..., 11.666667 , 11.666667 ,
          11.666667 ],
         [ 2.3333335,  2.3333335,  2.3333335, ..., 11.666667 , 11.666667 ,
          11.666667 ]], dtype=float32),
  )
# ---
# name: test_calculate_responsibility[Distribution.Uniform-dist_args0]
  Explanation(
    args=CausalArgs(
      all=False,
      alpha=0.2,
      analyze=False,
      batch=1,
      binary_threshold=None,
      blend=0.0,
      chunk_size=10,
      colour=200,
      concentrate=False,
      confidence_filter=0.0,
      config_location=None,
      data_location=None,
      db=None,
      distribution=<Distribution.Uniform: 2>,
      distribution_args=None,
      gpu=False,
      grid=False,
      heatmap=None,
      heatmap_colours='magma',
      info=True,
      insertion_step=100,
      iters=2,
      mark_segments=False,
      mask_value=0,
      means=None,
      min_box_size=10,
      mode=None,
      model=None,
      no_expansions=50,
      norm=255.0,
      output=None,
      path='tests/test_data/dog.jpg',
      processed=False,
      progress=True,
      queue_len=1,
      queue_style=<Queue.Area: 1>,
      raw=False,
      resize=False,
      search_limit=1000,
      seed=42,
      segmentation=False,
      shape=None,
      spatial_eta=0.2,
      spatial_radius=25,
      spotlight_eta=0.2,
      spotlight_size=20,
      spotlight_step=5,
      spotlights=10,
      stds=None,
      strategy=<Strategy.Spatial: 2>,
      surface=None,
      tree_depth=10,
      verbosity=0,
      weighted=False,
    ),
    data=Data: RGB, ['N', 3, 224, 224], 224, 224, 3, first
  	 Target:CLASS: 207, CONF: 0.20462, TARGET_CLASS: None, TARGET_CONFIDENCE: None,
    explanation=None,
    final_mask=None,
    maps={207: 4},
    run_stats=dict({
      'avg_box_size': 183.5625,
      'max_depth_reached': 5,
      'total_failing': 126,
      'total_passing': 84,
    }),
    target_map=array([[6.6666665, 6.6666665, 6.6666665, ..., 6.6666665, 6.6666665,
          6.6666665],
         [6.6666665, 6.6666665, 6.6666665, ..., 6.6666665, 6.6666665,
          6.6666665],
         [6.6666665, 6.6666665, 6.6666665, ..., 6.6666665, 6.6666665,
          6.6666665],
         ...,
         [5.5      , 5.5      , 5.5      , ..., 5.5      , 5.5      ,
          5.5      ],
         [5.5      , 5.5      , 5.5      , ..., 5.5      , 5.5      ,
          5.5      ],
         [5.5      , 5.5      , 5.5      , ..., 5.5      , 5.5      ,
          5.5      ]], dtype=float32),
  )
# ---
# name: test_preprocess
  Data: RGB, ['N', 3, 224, 224], 224, 224, 3, first
# ---
# name: test_preprocess_custom
  Data: RGB, ['N', 3, 224, 224], 224, 224, 3, first
# ---
# name: test_preprocess_npy
  Data: tabular, ['N', 3, 224, 224], None, None, None, None
# ---
# name: test_preprocess_spectral_mask_on_image_returns_warning
  Data: RGB, ['N', 3, 224, 224], 224, 224, 3, first
# ---
